### ISSUE TITLE
New Rule: `no-new-bigint`

### ISSUE THREAD

**Issue by sosukesuzuki** [eslint/eslint#16322]
### Rule details

Restricts calling `BigInt` with `new` like [`no-new-symbol`](https://eslint.org/docs/latest/rules/no-new-symbol)

### Related ECMAScript feature

[BigInt](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt)

### What type of rule is this?

Warns about a potential problem

### Example code

```js
// valid
const a = BigInt(9007199254740991);

// invalid
const b = new BigInt(9007199254740991);
```

### Why should this rule be in the core instead of a plugin?

`BigInt` is a core feature of ECMAScript, also ESLint have [`no-new-symbol`](https://eslint.org/docs/latest/rules/no-new-symbol) already.

### Participation

- [X] I am willing to submit a pull request to implement this rule.

**Comment by nzakas** [comment]
Maybe it would be better to update no-new-wrappers to include bigint?

https://eslint.org/docs/latest/rules/no-new-wrappers

**Comment by mdjermanovic** [comment]
A difference is that `new String(foo)`, `new Number(foo)` and `new Boolean(foo)` create wrapper objects, which `no-new-wrappers` rule intends to disallow, whereas `new BigInt(foo)` always throws `TypeError: BigInt is not a constructor`. If we follow the precedent of `no-new-symbol`, this looks more like a case for a new rule.

**Comment by nzakas** [comment]
Ah gotcha, read things a bit too fast.

We have said in the past that we didn’t want to add new rules that simply disallowed some syntax. I’m not sure it’s valuable to start going down the path of creating a new rule for each function that looks like a constructor.

**Comment by mdjermanovic** [comment]
> I’m not sure it’s valuable to start going down the path of creating a new rule for each function that looks like a constructor.

I agree. If we want to warn on `new BigInt`, then it might be better to add a rule that would warn on all globals that start with an uppercase letter but are not constructors (except those that are not functions, because they are included in [`no-obj-calls`](https://eslint.org/docs/latest/rules/no-obj-calls)). It would include `new Symbol`, so the `no-new-symbol` rule would be deprecated.

**Comment by nzakas** [comment]
I don’t feel like this type of error rises to the level of importance to be included in the core. At this point, with over 300 core rules, I’d like to be sure any new rule we add is “must-have.”

**Comment by mdjermanovic** [comment]
I wouldn't mind replacing (deprecating) `no-new-symbol` with a rule that would work the same, just checking a list of variable names - `["Symbol", "BigInt"]` at this point - instead of just [`"Symbol"`](https://github.com/eslint/eslint/blob/f6d57fb657c2f4e8e0140ad057da34c935482972/lib/rules/no-new-symbol.js#L35). Maintenance would be basically the same, and it would be easy to add new confusing globals that might appear in future specifications.

**Comment by nzakas** [comment]
I’m on board for that approach. 👍

**Comment by sosukesuzuki** [comment]
Thank you. What should we name it: `no-constructor-call`? `no-new-call`?

**Comment by mdjermanovic** [comment]
Maybe something like `no-new-function` or `no-new-nonconstructor`?

**Comment by sosukesuzuki** [comment]
I prefer `no-new-nonconstructor`.
